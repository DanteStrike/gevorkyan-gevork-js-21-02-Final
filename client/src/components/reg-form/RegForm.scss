@import "src/assets/variables";

.reg-form {
  width: 380px;

  &__gender .ant-form-item-control-input-content {
    display: flex;
  }

  &__gender-container {
    margin: 0 auto;
  }

  .ant-radio-wrapper {
    color: var(--primary-font-color);
  }

  &__date-picker {
    width: 100%;
  }

  &__last {
    margin-bottom: 0;
  }

  &__link {
    color: var(--secondory-font-color);
    text-align: center;
  }

  .ant-form-item-label {
    padding: 0;
    font-weight: 700;
  }

  .ant-form-item-label > label {
    color: var(--primary-font-color);
  }
}

@media (max-width: $tabletWidth) {
  .reg-form {
    width: 85%;
  }
}
