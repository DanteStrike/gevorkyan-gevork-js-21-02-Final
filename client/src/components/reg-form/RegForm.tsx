import React from 'react';
import './RegForm.scss';
import {Moment} from 'moment';
import {Form, Input, DatePicker, Radio} from 'antd';
import Button from '../submit-button/SubmitButton';
import CustomLink from '../custom-link/CustomLink';
import {IUserRegistration} from '../../types';
import {RoutePath} from '../../enums';
import {DataUtils, ValidateUtils} from '../../utils';
import useAppTranslation from '../../hooks/use-app-translation';

interface IRegFormProps {
  loading?: boolean;
  onSubmit?: (data: IUserRegistration) => void;
}
type RegFormValuesType = Omit<IUserRegistration, 'firstName' | 'lastName' | 'dateOfBirth'> & {
  name: string;
  dateOfBirth: Moment;
};

const {Item} = Form;

function RegForm({loading, onSubmit = () => {}}: IRegFormProps) {
  const {t} = useAppTranslation(`regForm`);
  const [form] = Form.useForm();

  const handleFormFinish = (filedValues: RegFormValuesType) => {
    const {name, dateOfBirth, ...values} = filedValues;
    const normName = DataUtils.normalizeName(name);
    const userReg: IUserRegistration = {
      ...normName,
      ...values,
      dateOfBirth: dateOfBirth.toISOString(),
    };
    onSubmit(userReg);
  };

  return (
    <Form
      className="reg-form"
      form={form}
      name="registration"
      size="middle"
      layout="vertical"
      validateTrigger="onSubmit"
      onFinish={handleFormFinish}
      initialValues={{
        name: ``,
        gender: `male`,
        dateOfBirth: ``,
        email: ``,
        phone: ``,
      }}
    >
      <Item
        name="name"
        label={t(`name.title`)}
        rules={[ValidateUtils.requireValidatorName, ValidateUtils.userNameValidator]}
      >
        <Input placeholder={t(`name.placeholder`)} />
      </Item>

      <Item className="reg-form__gender" name="gender" label={t(`gender.title`)}>
        <Radio.Group className="reg-form__gender-container">
          <Radio value="male">{t(`gender.options.male`)}</Radio>
          <Radio value="female">{t(`gender.options.female`)}</Radio>
          <Radio value="other">{t(`gender.options.other`)}</Radio>
        </Radio.Group>
      </Item>

      <Item
        name="dateOfBirth"
        label={t(`dateOfBirth.title`)}
        rules={[ValidateUtils.requireValidatorBirth, ValidateUtils.sevenYearsOldValidator]}
      >
        <DatePicker
          className="reg-form__date-picker"
          placeholder={t(`dateOfBirth.placeholder`)}
          disabledDate={ValidateUtils.disableDateOverCurrent}
        />
      </Item>

      <Item
        name="email"
        label={t(`email.title`)}
        rules={[ValidateUtils.requireValidatorEmail, ValidateUtils.emailFormatValidator()]}
      >
        <Input placeholder={t(`email.placeholder`)} />
      </Item>

      <Item
        name="phone"
        label={t(`phone.title`)}
        rules={[ValidateUtils.requireValidatorPhone, ValidateUtils.phoneFormatValidator()]}
      >
        <Input placeholder={t(`phone.placeholder`)} />
      </Item>

      <Item className="reg-form__last">
        <Button loading={loading}>{t(`register`)}</Button>
        <CustomLink className="reg-form__link" to={RoutePath.LOGIN}>
          {t(`message`)}
        </CustomLink>
      </Item>
    </Form>
  );
}

RegForm.defaultProps = {
  loading: false,
  onSubmit: () => {},
};

export default React.memo(RegForm);
