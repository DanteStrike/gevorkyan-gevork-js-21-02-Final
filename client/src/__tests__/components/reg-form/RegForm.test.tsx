import {mount, shallow} from 'enzyme';
import React from 'react';
import {Router} from 'react-router-dom';
import {createMemoryHistory} from 'history';
import moment from 'moment';
import RegForm from '../../../components/reg-form/RegForm';
import sleep from '../../mocks/utils';

describe(`Component RegForm should render correctly`, () => {
  it(`snapshot default test`, () => {
    const wrapper = shallow(<RegForm />);
    expect(wrapper).toMatchSnapshot();
  });

  it(`snapshot loading test`, () => {
    const wrapper = shallow(<RegForm loading />);
    expect(wrapper).toMatchSnapshot();
  });
});

describe(`Component RegForm should work correctly`, () => {
  it(`call submit test`, () => {
    const history = createMemoryHistory();
    const submitHandler = jest.fn(() => {});
    const wrapper = mount(
      <Router history={history}>
        <RegForm onSubmit={submitHandler} />
      </Router>
    );

    const mockData = {
      name: `full user name`,
      gender: `female`,
      dateOfBirth: moment().subtract(18, 'years'),
      email: `email@mail.ru`,
      phone: `+79999999999`,
    };

    wrapper.children().children().prop(`form`).setFieldsValue(mockData);
    wrapper.update();
    wrapper.find('form').simulate('submit');

    return sleep().then(() => {
      expect(submitHandler).toHaveBeenCalledTimes(1);
    });
  });
});
