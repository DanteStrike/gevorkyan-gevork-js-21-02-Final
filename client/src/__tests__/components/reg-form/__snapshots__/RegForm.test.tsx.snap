// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component RegForm should render correctly snapshot default test 1`] = `
<ForwardRef(InternalForm)
  className="reg-form"
  form={
    Object {
      "__INTERNAL__": Object {
        "itemRef": [Function],
      },
      "getFieldError": [Function],
      "getFieldInstance": [Function],
      "getFieldValue": [Function],
      "getFieldsError": [Function],
      "getFieldsValue": [Function],
      "getInternalHooks": [Function],
      "isFieldTouched": [Function],
      "isFieldValidating": [Function],
      "isFieldsTouched": [Function],
      "isFieldsValidating": [Function],
      "resetFields": [Function],
      "scrollToField": [Function],
      "setFields": [Function],
      "setFieldsValue": [Function],
      "submit": [Function],
      "validateFields": [Function],
    }
  }
  initialValues={
    Object {
      "dateOfBirth": "",
      "email": "",
      "gender": "male",
      "name": "",
      "phone": "",
    }
  }
  layout="vertical"
  name="registration"
  onFinish={[Function]}
  size="middle"
  validateTrigger="onSubmit"
>
  <FormItem
    label="Full name:"
    name="name"
    rules={
      Array [
        [Function],
        [Function],
      ]
    }
  >
    <Input
      placeholder="Input full name"
      type="text"
    />
  </FormItem>
  <FormItem
    className="reg-form__gender"
    label="Gender:"
    name="gender"
  >
    <Memo(ForwardRef)
      className="reg-form__gender-container"
    >
      <Radio
        type="radio"
        value="male"
      >
        Male
      </Radio>
      <Radio
        type="radio"
        value="female"
      >
        Female
      </Radio>
      <Radio
        type="radio"
        value="other"
      >
        Other
      </Radio>
    </Memo(ForwardRef)>
  </FormItem>
  <FormItem
    label="Date of birth:"
    name="dateOfBirth"
    rules={
      Array [
        [Function],
        [Function],
      ]
    }
  >
    <Picker
      className="reg-form__date-picker"
      disabledDate={[Function]}
      placeholder="Choose date"
    />
  </FormItem>
  <FormItem
    label="E-mail:"
    name="email"
    rules={
      Array [
        [Function],
        Object {
          "message": "Input correct email",
          "type": "email",
        },
      ]
    }
  >
    <Input
      placeholder="anonim@gmail.com"
      type="text"
    />
  </FormItem>
  <FormItem
    label="Phone:"
    name="phone"
    rules={
      Array [
        [Function],
        Object {
          "message": "Input correct phone",
          "pattern": /\\^\\\\\\+\\?7\\(\\\\d\\{10\\}\\)\\$/,
        },
      ]
    }
  >
    <Input
      placeholder="+79995553322"
      type="text"
    />
  </FormItem>
  <FormItem
    className="reg-form__last"
  >
    <Memo(SubmitButton)
      loading={false}
    >
      Register
    </Memo(SubmitButton)>
    <Memo(CustomLink)
      className="reg-form__link"
      to="/login"
    >
      Do you have account? Login
    </Memo(CustomLink)>
  </FormItem>
</ForwardRef(InternalForm)>
`;

exports[`Component RegForm should render correctly snapshot loading test 1`] = `
<ForwardRef(InternalForm)
  className="reg-form"
  form={
    Object {
      "__INTERNAL__": Object {
        "itemRef": [Function],
      },
      "getFieldError": [Function],
      "getFieldInstance": [Function],
      "getFieldValue": [Function],
      "getFieldsError": [Function],
      "getFieldsValue": [Function],
      "getInternalHooks": [Function],
      "isFieldTouched": [Function],
      "isFieldValidating": [Function],
      "isFieldsTouched": [Function],
      "isFieldsValidating": [Function],
      "resetFields": [Function],
      "scrollToField": [Function],
      "setFields": [Function],
      "setFieldsValue": [Function],
      "submit": [Function],
      "validateFields": [Function],
    }
  }
  initialValues={
    Object {
      "dateOfBirth": "",
      "email": "",
      "gender": "male",
      "name": "",
      "phone": "",
    }
  }
  layout="vertical"
  name="registration"
  onFinish={[Function]}
  size="middle"
  validateTrigger="onSubmit"
>
  <FormItem
    label="Full name:"
    name="name"
    rules={
      Array [
        [Function],
        [Function],
      ]
    }
  >
    <Input
      placeholder="Input full name"
      type="text"
    />
  </FormItem>
  <FormItem
    className="reg-form__gender"
    label="Gender:"
    name="gender"
  >
    <Memo(ForwardRef)
      className="reg-form__gender-container"
    >
      <Radio
        type="radio"
        value="male"
      >
        Male
      </Radio>
      <Radio
        type="radio"
        value="female"
      >
        Female
      </Radio>
      <Radio
        type="radio"
        value="other"
      >
        Other
      </Radio>
    </Memo(ForwardRef)>
  </FormItem>
  <FormItem
    label="Date of birth:"
    name="dateOfBirth"
    rules={
      Array [
        [Function],
        [Function],
      ]
    }
  >
    <Picker
      className="reg-form__date-picker"
      disabledDate={[Function]}
      placeholder="Choose date"
    />
  </FormItem>
  <FormItem
    label="E-mail:"
    name="email"
    rules={
      Array [
        [Function],
        Object {
          "message": "Input correct email",
          "type": "email",
        },
      ]
    }
  >
    <Input
      placeholder="anonim@gmail.com"
      type="text"
    />
  </FormItem>
  <FormItem
    label="Phone:"
    name="phone"
    rules={
      Array [
        [Function],
        Object {
          "message": "Input correct phone",
          "pattern": /\\^\\\\\\+\\?7\\(\\\\d\\{10\\}\\)\\$/,
        },
      ]
    }
  >
    <Input
      placeholder="+79995553322"
      type="text"
    />
  </FormItem>
  <FormItem
    className="reg-form__last"
  >
    <Memo(SubmitButton)
      loading={true}
    >
      Register
    </Memo(SubmitButton)>
    <Memo(CustomLink)
      className="reg-form__link"
      to="/login"
    >
      Do you have account? Login
    </Memo(CustomLink)>
  </FormItem>
</ForwardRef(InternalForm)>
`;
